src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,936 @@
+
+    @staticmethod
+    def empty(nblocks, dtype=np.float64):
+        """
+        Instantiate an empty BlkDiagMatrix with `nblocks`, where each
+        data block is initially None with size (0,0).
+
+        This is used for incrementally building up a BlkDiagMatrix, by
+        using nblocks=0 in conjunction with `append` method or
+        situations where blocks are immediately assigned in a loop,
+        such as in a `copy`.
+
+        :param nblocks: Number of diagonal matrix blocks.
+        :return BlkDiagMatrix instance where each block is None.
+        """
+
+        # Empty partition has block dims of zero until they are assigned
+        partition = [(0, 0)] * nblocks
+
+        return BlkDiagMatrix(partition, dtype=dtype)
+
+    def append(self, blk):
+        """
+        Append `blk` to `self`. Used to incrementally build up a BlkDiagMatrix
+        instance where the number of blocks and/or shapes are derived
+        incrementally.
+
+        :param blk: Block to append (ndarray).
+        """
+
+        self.data.append(blk)
+        self.nblocks += 1
+        self.reset_cache()
+
+    def __repr__(self):
+        """
+        String represention describing instance.
+        """
+        return "BlkDiagMatrix({}, {})".format(
+            repr(self.nblocks), repr(self.dtype))
+
+    def copy(self):
+        """
+        Returns new BlkDiagMatrix which is a copy of `self`.
+
+        :return BlkDiagMatrix like self
+        """
+
+        C = BlkDiagMatrix(self.partition, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            C[i] = self[i].copy()
+
+        return C
+
+    # Manually overload list methods,
+    #   This is just for syntax which allows us to reference self[i] etc
+    #     instead of writing self.data all the time. You may use either.
+    def __getitem__(self, key):
+        """
+        Convenience wrapper, getter on self.data.
+        """
+
+        return self.data[key]
+
+    def __setitem__(self, key, value):
+        """
+        Convenience wrapper, setter on self.data.
+        """
+
+        BlkDiagMatrix.check_square(value.shape)
+        self.data[key] = value
+        self.reset_cache()
+
+    def __len__(self):
+        """
+        Convenience function for getting nblocks.
+        """
+
+        return self.nblocks
+
+    def _is_scalar_type(self, x):
+        """
+        Internal helper function checking scalar-ness for elementwise ops.
+
+        Essentially we are checking for a single numeric object, as opposed to
+        something like an `ndarray` or `BlkDiagMatrix`. We do this by
+        checking `numpy.isscalar(x)`.
+
+        In the future this check may require extension to include ASPIRE or
+        other third party types beyond what is provided by numpy, so we
+        implement it now as a class method.
+
+        :param x: Value to check
+
+        :return: bool.
+        """
+
+        return np.isscalar(x)
+
+    @staticmethod
```

Comment:
Can we group all these static methods together?

Reply:
Yes, I moved them all the the end of the class.